src/index.ts

```
@@ -140,94 +142,149 @@ export class Engine {
+                console.error(logStream.input._buffer);
+            });
+            // Connect log hooks:
+            child.stdout.pipe(logStream.input);
+            child.stderr.pipe(logStream.input);
+
+            // Feed config into process:
+            child.stdin.write(childConfigJson);
 
-                const childConfigJson = JSON.stringify(childConfig) + '\n';
-                const spawnChild = () => {
-                    const child = spawn(binaryPath, ['-config=stdin']);
-                    this.child = child;
-
-                    const logStream = StreamJsonObjects.make();
-                    logStream.output.on('data', (logData: any) => {
-                        const logRecord = logData.value;
-                        console.log(JSON.stringify({proxy: logRecord}));
-                    });
-
-                    // Connect log hooks:
-                    child.stdout.pipe(logStream.input);
-                    child.stderr.pipe(logStream.input);
-
-                    // Feed config into process:
-                    child.stdin.write(childConfigJson);
-
-                    // Connect shutdown hooks:
-                    child.on('exit', (code, signal) => {
-                        if (!this.running) {
-                            // It's not an error if we think it's our fault.
-                            return;
-                        }
-
-                        if (code != null) {
-                            console.error(`Engine crashed unexpectedly with code: ${code}`);
-                        }
-                        if (signal != null) {
-                            console.error(`Engine was killed unexpectedly by signal: ${signal}`);
-                        }
-                        spawnChild();
-                    });
-                };
+            // Connect shutdown hooks:
+            child.on('exit', (code, signal) => {
+                // Wipe the URI, so middleware doesn't route to dead process:
```

If I understand the code correctly, while it's restarting middlewareParams.uri will be empty and the middleware will just be ignored. Would it not make more sense to buffer the request until engineproxy is back? What's the guarantee that you'll properly process an intended-for-engineproxy request if it doesn't get sent to engineproxy?

---

Yep; that's the intent.

I think if the proxy binary crashes, it's best to keep engine out of the middleware.
I'd rather my site be up with a few cache misses/missing traces, than my site be down/lagged because of an issue in the proxy binary (that is closed source and I can't debug).

---

OK. I'd love it if somebody with more holistic engine understanding than me had eyes on this. This is making the strong assumption that Engine adds no "non-essential" features.  eg if we had a security layer in engineproxy, we wouldn't want to sometimes bypass it. I'm not even sure if we're already at that point with persisted queries.

cc @pcarrier @timbotnik @daniman @josephglanville

---

wrt persisted queries, if the client gets an error other than "PersistedQueryNotFound" it will turn off persisted queries for that session (assuming the server responds with a GraphQL parse error or similar in this case).  This of course means every new session will have an extra round-trip and the higher network costs associated.

While we probably can (at the moment) limp along w/o Engine, my feeling is that failing hard / fast is preferred since the behavior is fairly undefined and perhaps hard to debug otherwise.  Also buffering the requests sounds hairy as well, could we just respond with a 5xx while the proxy is down?

---

I'm going to bow out of this decision and leave it up to Tim and the other mentioned folks who are more deeply involved with Engine.

---

The logic behind "we should passthrough requests when the proxy has crashed" is https://github.com/apollographql/GitHunt-API/issues/264.

Open source project wants to include engine in their code base, but doesn't want to commit the API key.
They _could_ `if (ENGINE_API_KEY) { engine.start(); }`, but that was the use case.

---

But there is a PR https://github.com/apollographql/GitHunt-API/pull/265/files to fix it with an if statement, which seems like a good-enough solution for their use case.  I definitely feel like we should respond with errors when the proxy _has been started_ and _while it is down_ (this would be the behavior of non-sidecar operation) but I also would really like to get this PR merged which if I understand correctly is blocking release of persisted queries.  @thepwagner let's follow up quickly with another PR where we can get some other opinions on the merits of passthru vs. fail hard vs. buffer.

---

Noice PR find; I only saw the internal facing request in YouTrack - then dug up the public issue for context here.

Breadcrumbed to https://github.com/apollographql/apollo-engine-js/issues/61